.devcontainer/Dockerfile

```diff
@@ -0,0 +1,57 @@
+FROM ubuntu:focal
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN echo deb http://dk.archive.ubuntu.com/ubuntu/ xenial main >> /etc/apt/sources.list \
+    && echo deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe >> /etc/apt/sources.list \
+    && dpkg --add-architecture i386 \
+    && apt update
+
+# Install common build tools
+RUN apt install -y --no-install-recommends ca-certificates curl zip unzip tar git ninja-build clang-10 pkg-config
+
+# Setup CMake
+RUN cd /usr/local \
+    && (curl -sL https://github.com/Kitware/CMake/releases/download/v3.23.2/cmake-3.23.2-linux-x86_64.sh) \
+    > /tmp/cmakeinstall.sh \
+    && bash /tmp/cmakeinstall.sh --skip-license
+
+# Setup Linux toolchain
+RUN apt install -y --no-install-recommends g++-4.8-multilib \
+    && mkdir -p /usr/gcc-4.8-clang-toolchain/lib/gcc/x86_64-linux-gnu \
+    && cd /usr/gcc-4.8-clang-toolchain \
+    && ln -s /usr/include include \
+    && ln -s /usr/bin bin \
+    && cd lib/gcc/x86_64-linux-gnu \
+    && ln -s /usr/lib/gcc/x86_64-linux-gnu/4.8 4.8
+
+# Setup Windows toolchain
+RUN apt install -y --no-install-recommends g++-mingw-w64-i686
+
+# Setup powershell (required for Windows build when vcpkg is involved)
+RUN (curl -sL https://github.com/PowerShell/PowerShell/releases/download/v7.2.4/powershell-lts_7.2.4-1.deb_amd64.deb) \
+    > /tmp/pwsh.deb \
+    && (dpkg -i /tmp/pwsh.deb; apt install -y --no-install-recommends -f)
+
+# TODO: vcpkg-fy
+# RUN apt install -y --no-install-recommends lua5.3 luarocks liblua5.3-dev
+
+# Setup vcpkg
+RUN cd / \
+    && git clone https://github.com/microsoft/vcpkg \
+    && cd /vcpkg \
+    && ./bootstrap-vcpkg.sh \
+    && ln -s /vcpkg/vcpkg /usr/bin/vcpkg
+COPY triplets/* /vcpkg/triplets/
+COPY toolchains/* /vcpkg/scripts/toolchains/
+
+# Workarounds for boost
+RUN ln -s /usr/bin/clang-10 /usr/bin/clang && ln -s /usr/bin/clang++-10 /usr/bin/clang++
```

Comment:
> Was it not possible to simply install the boost-dev & sdl2-dev libraries so that the headers are in the proper include location already?
There's no apt package of them cross-compiled for Windows. This is why vcpkg is used here, it can build pretty much any dependencies for any compiler setups.

I believe boost can be used from Linux distro if we're careful with using only header-only features, but that's a potential tripwire anyway.

> Why were these workarounds for Boost & SDL2 needed?

* Boost: its build script (b2) doesn't fully recognize setups from CMake so it cannot find clang which is named `clang-10`.
/ SDL2: I'm not sure what's happening on this one. It tries to find `asm/errno.h` during compilation which doesn't exist but in `asm-generic` instead.

Reply:
That's interesting because we don't build boost, from my understanding boost is just headers?

Reply:
sdl2 also isn't used by the windows build at all it's only used on Linux right now.

It would be nice if it was but you'd still only want the headers and then distribute a `libsdl2.dll` alongside the windows build unless it was statically linked.

Reply:
There's something you have to set on debian/Ubuntu to set the default compiler name for clang so that "clang" resolves to clang-10.

A link can probably do it but there was some specific script that I think was mentioned in the console output when I first installed clang with apt. Like "do this to setup this clang as the default clang" and it spat out a couple lines to run in the terminal.

Reply:
> That's interesting because we don't build boost, from my understanding boost is just headers?

[.cbp file has some lines related to linking boost libs](https://github.com/FTL-Hyperspace/FTL-Hyperspace/blob/0c63d385784ed36538dff4ef6e560fbab14d1e96/FTL_HyperspaceGCC.cbp#L40-L76). I'm not sure if all of them are redundant.

boost-math is a [buildable library](https://www.boost.org/doc/libs/1_79_0/libs/math/doc/html/math_toolkit/building.html) though I doubt if `boost::math::isnan()` requires building. Also some of the other boost components have deps that needs to be built when used with vcpkg.

If we can strictly limit the usage to only header-only parts of the boost we can pull Linux build from apt then use headers from it but that's a potential tripwire anyway.

> sdl2 also isn't used by the windows build at all it's only used on Linux right now.
>
>It would be nice if it was but you'd still only want the headers and then distribute a libsdl2.dll alongside the windows build unless it was statically linked.

[Here's why I thought it's a good idea to add SDL into Windows build.](https://github.com/FTL-Hyperspace/FTL-Hyperspace/blob/master/NotesToWindowsDevs.md#things-not-yet-reimplemented-but-planned)

And I thought everything was configured to be statically linked for Windows?

> There's something you have to set on debian/Ubuntu to set the default compiler name for clang so that "clang" resolves to clang-10.
>
> A link can probably do it but there was some specific script that I think was mentioned in the console output when I first installed clang with apt. Like "do this to setup this clang as the default clang" and it spat out a couple lines to run in the terminal.

Hmm. I did run `update-alternatives` for that purpose but it didn't list clang:
```
root@a592d2b0000a:/workspaces/FTL-Hyperspace# update-alternatives --get-selections
awk                            auto     /usr/bin/mawk
i686-w64-mingw32-g++           auto     /usr/bin/i686-w64-mingw32-g++-win32
i686-w64-mingw32-gcc           auto     /usr/bin/i686-w64-mingw32-gcc-win32
pager                          auto     /bin/more
rmt                            auto     /usr/sbin/rmt-tar
w                              auto     /usr/bin/w.procps
```

Reply:
> [.cbp file has some lines related to linking boost libs in native Windows build](https://github.com/FTL-Hyperspace/FTL-Hyperspace/blob/0c63d385784ed36538dff4ef6e560fbab14d1e96/FTL_HyperspaceGCC.cbp#L40-L76). I'm not sure if all of them are redundant.

Yeah on Windows they statically link it, Linux dynamically links everything because dynamic linking is better and is the standard in the Linux ecosystem.
Boost is *basically* available on every Linux system, while it isn't a guarantee so many other programs use it that the library is already there by the time you've got a system capable of playing video games.
So on Linux we never need to build boost, just like we never need to build SDL2 but users may have to install libboost or libsdl2 to run Hyperspace if their system is missing them. It might be *nice* to distribute a `libboost.so` and `libsdl2.so` alongside the Hyperspace so files (and soon a `liblua.so`) to support the Steam Deck (assuming they're not already preinstalled on Steam Deck, pretty much every other Linux OS people can just install the library anyways with their package manager) but we don't need to (or want to) build them every time, nor is it important that they're built with the same version of clang or whatever, they're dynamic libraries and can (and maybe should? not sure) be built with the system's default GCC (or we just rely on them installing the library from their package manager).

Windows is awful though with how the libraries work, so boost either needs to be included statically (that's why there's a statically compiled `a` file I think) or it needs to be dynamically linked and included as a DLL alongside Hyperspace.dll.
Now I don't know for static libraries if they need to be built with the same compiler or not.

> [Here's why I thought it's a good idea to add SDL into Windows build.](https://github.com/FTL-Hyperspace/FTL-Hyperspace/blob/master/NotesToWindowsDevs.md#things-not-yet-reimplemented-but-planned)

Yeah it's a good point, if we have a reliable way of including SDL in the Windows build then I can use the nice message box; however, since that original comment we did change our code from many separate calls to the message box to one utility function that uses the message box and right now that's the only one that has a preprocessor statement to switch between SDL on Linux & the normal Windows MessageBoxA method on Windows. So while it might be nice, it doesn't currently serve a huge benefit. **If you want to do it, it's more power to you** but we don't *need* it as annoyingly as when I originally wrote that comment.

> And I thought everything was configured to be statically linked for Windows?

It is, except for Lua 5.3 (which is only in the `develop` branch right now and not yet in `master` as we haven't published a build that includes it yet), Lua isn't allowed to be statically linked for some reason. I didn't investigate it more but Lua basically says "don't" and so Lua is dynamically linked and will require a `liblua.dll` or whatever it's called (I don't remember what the default name is) to be next to `Hyperspace.dll` when running. (or somewhere else on the system path but that's far too complicated for Windows users)... On Linux it will of course just mean install `lua5.3` or maybe `5.4` from your package manager (but for the Steam Deck it might still be useful to distribute a liblua5.3.so although that might fit as a separate docker container that does that build from Lua's source and we just copy the file into the zip we build).

> > There's something you have to set on debian/Ubuntu to set the default compiler name for clang so that "clang" resolves to clang-10.
> > A link can probably do it but there was some specific script that I think was mentioned in the console output when I first installed clang with apt. Like "do this to setup this clang as the default clang" and it spat out a couple lines to run in the terminal.
> 
> Hmm. I did try running `update-alternatives` for that purpose but it didn't list clang:

Weird because that sounds familiar but maybe that was from when I was installing GCC, however, when I run `clang -v` my default is 10 even though I do have multiple versions of clang installed
```sh
 ~ clang -v
clang version 10.0.0-4ubuntu1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
...
```
Note however, `clang` is not in my `update-alternatives --get-selections` options so maybe it did just tell me to make a `ln` or something after the install. I don't really want to re-install it right now to find out.